service/s3/internal/customizations/update_endpoint.go

```diff
@@ -7,33 +7,33 @@ import (
 	"regexp"
 	"strings"
 
-	"github.com/awslabs/smithy-go"
 	"github.com/awslabs/smithy-go/middleware"
 	"github.com/awslabs/smithy-go/transport/http"
 )
 
 // UpdateEndpointOptions provides the options for the UpdateEndpoint middleware setup.
 type UpdateEndpointOptions struct {
-	UsePathStyle  bool
-	UseAccelerate bool
-
-	Region string
+	// functional pointer to fetch bucket name from provided input.
+	// The function is intended to take an input value, and
+	// return a string pointer to value of string, and bool if
+	// input has no bucket member.
+	GetBucketFromInput func(interface{}) (*string, bool)
+
+	// use path style
+	UsePathStyle bool
 }
 
 // UpdateEndpoint adds the middleware to the middleware stack based on the UpdateEndpointOptions.
 func UpdateEndpoint(stack *middleware.Stack, options UpdateEndpointOptions) {
 	stack.Serialize.Insert(&updateEndpointMiddleware{
-		usePathStyle:  options.UsePathStyle,
-		useAccelerate: options.UseAccelerate,
-		region:        options.Region,
+		getBucketFromInput: options.GetBucketFromInput,
+		usePathStyle:       options.UsePathStyle,
 	}, "OperationSerializer", middleware.After)
 }
```

Comment:
Should this be inserted after the endpoint resolver?  or does endpoint resolving happen before operation serializing?

Reply:
Endpoint resolving happens before operation serializing